container/containerd/client.go

```diff
@@ -28,20 +28,24 @@ import (
 	"github.com/containerd/containerd/errdefs"
 	"github.com/containerd/containerd/pkg/dialer"
 	ptypes "github.com/gogo/protobuf/types"
+	"github.com/google/cadvisor/container/common"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
```

Comment:
I'm not sure if we can depend on k8s repos since cadvisor itself is imported into k8s which can result in an import loop. 

See https://github.com/google/cadvisor/pull/2726 and https://github.com/google/cadvisor/pull/2726#issuecomment-725363880

Do we need to import criapi here? Can we just talk to containerd directly?

Reply:
It's ok ,  this package is just pb files.

And there are two way to get the diskusage metrics : 
* call snapshots.Usage()
* call cri.ContainerStats()

They are both grpc interface implemented by containerd , the difference is that `snapshots.Usage` didn‘t cache any thing, while containerd cri server will sync the disk usage metrics periodically, see [snapshotsSyncer.sync](https://github.com/containerd/containerd/blob/main/pkg/cri/server/snapshots.go#L71) ,in  order to avoid collecting disk usage metric repeatedly , we should use the cri method.

Reply:
I see, thanks for the background. Unfortunately, I don't think the fact that this package is just pb files matters; I think it is general issue with dependency management in k8s => 

as @dims mentioned [here](https://github.com/google/cadvisor/pull/2726#issuecomment-725363880)

> "third party software can use staging modules IFF the third party software is not vendored back into k/k :)"

"k8s.io/cri-api/" is staging module and cadvisor itself is vendored back into k/k. 

@dims, please keep me honest here, but I don't think anything has changed here.

Reply:
Will be able to switch to `snapshots.Usage()` instead? Can we maybe cache the results internally / only collect them during some periodic house keeping period?

Reply:
If we just consider this feature in cadvisor independently , we can use `snapshots.Usage`, but for the sake of performance ,we should use `cri.ContainerStats` instead.  there is a picture which shows the relationshipt between cadvisor ,kubelet  and containerd in this situation :

![diskmetrics](https://user-images.githubusercontent.com/8213509/123188767-2b9d9b00-d4cf-11eb-80a7-ce329da0c4d2.jpg)

If we use `snapshots.Usage` in cadvisor, it will cause unnecessary performance loss.

Reply:
I didn't find an another equivalent  library , can I copy that stuff over to cadvisor ？

Reply:
I chatted with @dims offline and confirmed that importing `k8s.io/cri-api/` will result in circular dependency, so unfortunately we simply won't be able to do that.  

Do you have any idea for alternative idea? I understand `snapshots.Usage` is not cached, is there some other idea how to mitigate that? Can we maybe cache it internally? Or use/introduce some other containerd API?

Reply:
Maybe we could use `snapshots.Usage`  and cache it internally.  And at the same time , modify kubelet , do not collect diskUsageMetrics by cadvisor  when it's containerd, and use the data returned by [runtimeService.ListContainerStats](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go#L148).

But I don't known if it's ok . And I have sent an email to @Random-Liu  (who is familiar with this)  for more suggestion.

Reply:
That sounds reasonable to me. Is the idea then that cAdvisor standalone would use `snapshots.Usage` but when used inside kubelet, we would disable using `snapshots.Usage` in cadvisor and instead rely on `runtimeService.ListContainerStats`. Feel free to cc me on the email: my mail is porterdavid@google.com

Reply:
> I chatted with @dims offline and confirmed that importing k8s.io/cri-api/ will result in circular dependency, so unfortunately we simply won't be able to do that.

Hm, I don't quite understand. The import dependency is at package level right? As long as `k8s.io/cri-api` doesn't directly/indirectly depend any part cadvisor repo, it seems safe to me.

The only issue I can imagine is k8s and cadvisor depends on different versions of CRI, in that case, it seems to be resolvable.

We already have this kind of dependency today IIRC: both cadvisor and kubelet depend on the same docker client library, and containerd libraries. CRI is just yet another dependency like this, and an easier to manage one because kubelet, cadvisor and CRI are all heavily maintained by K8s maintainers.

Reply:
@Random-Liu please have a chat with @liggitt and fill us back in here.

Reply:
> The import dependency is at package level not repo level right?

no, it's at the module level.

> As long as k8s.io/cri-api doesn't directly/indirectly depend on any part of cadvisor repo, it seems safe to me.

k8s.io/cri-api is published from k8s.io/kubernetes (canonical location at https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/cri-api)

If a dependency vendored into k8s.io/kubernetes depends on something published out of k8s.io/kubernetes, then it creates gridlock, where a change in k8s.io/cri-api could require a new level of cadvisor, which cannot be made simultaneously.

Reply:
Hm, but the `github.com/containerd/containerd` module depends on `k8s.io/cri-api`, but also being vendored into `k8s.io/kubernetes`. How does that work?

Reply:
The packages we vendor from github.com/containerd/containerd don't include the k8s.io/cri-api reference, and the version k/k currently vendors (v1.4.4) doesn't have a go module, (https://github.com/containerd/containerd/blob/v1.4.4/go.mod) so our vendor scripts are unaware of the circular dependency. If they were, they would forbid it.

Reply:
The github.com/containerd/containerd circular dependency back on a k/k staging module is a timebomb that will keep us from picking up newer versions :-/

Reply:
I believe what containerd does is correct. We should make sure our CRI API is safe to import for other projects in the community. :/

Is it possible to define a dedicated module for `k8s.io/cri-api/pkg/apis/runtime`? It is just api definition without much extra dependency.

Reply:
Using CRI method is the better way ,but at current circular dependency is also a  real problem...  How about using `snapshots.Usage` for the time being ，and when `k8s.io/cri-api` is safe to import ,then switch to CRI?

Reply:
> How about using snapshots.Usage for the time being ，and when k8s.io/cri-api is safe to import ,then switch to CRI?

This seems like the best (only) option for now so I agree -- we should do that. However, we should be careful to cache the `snapshots.Usage` info internally in cAdvisor since it's not cached as you mentioned earlier.

Reply:
@bobbypage I have changed it to `snapshots.Usage`

Reply:
Thanks @yyrdl for the update! 

One thing I would like to understand before proceeding is the best path forward to integrate these changes back into kubelet. As I understand with this PR merged and cAdvisor vendored back into kubelet the story will be the following:

1. If using CRI runtime, kubelet will use `ListContainerStats` CRI method which containerd will use containerd's internal [`SnapshotSyncer`](https://github.com/containerd/containerd/blob/559f9a5b90073dc12005d1a611c97007db0be131/pkg/cri/server/service.go#L196-L200) which will internally call `Usage` on the snapshot [periodically](https://github.com/containerd/containerd/blob/559f9a5b90073dc12005d1a611c97007db0be131/pkg/cri/server/snapshots.go#L101) and cache the results. It looks like by default these stats are refreshed every [10 seconds] by containerd (https://github.com/containerd/containerd/blob/559f9a5b90073dc12005d1a611c97007db0be131/pkg/cri/config/config_unix.go#L95)
2. With this PR included, cAdvisor (vendored into kubelet) will also call `snapshots.Usage` (which is an expensive call). That value is not cached / shared with the CRI `SnapshotSyncer` method. How often will this happen (is this every house keeping period)?

as background, looks like `snapshot.Usage` implementation at least for [overlay snapshotter](https://github.com/containerd/containerd/blob/main/snapshots/overlay/overlay.go#L174) requires [walking the filesystem](https://github.com/containerd/containerd/blob/559f9a5b90073dc12005d1a611c97007db0be131/vendor/github.com/containerd/continuity/fs/du_unix.go#L51-L87) counting inodes and size by calling `stat` repeatedly  

As you mentioned previously, with this approach it sounds like we will collect the `snapshots.Usage` data twice from kubelet. So it sounds like we have the following options:

1. Collect the data twice (con: has potential perf impact -- i'm not sure how much -- this will be good to understand. Do you have any idea?)
2. Disable cAdvisor in kubelet from collecting container disk usage metrics. (pro: we avoid perf impact, con: kubelet /metrics/cadvisor endpoint will not report these metrics; they will only be available for users who run cAdvisor standalone/daemonset).

Do you have any thoughts on options there and perf impact? I think we should have clear idea how we integrate these changes back into kubelet before proceeding.

Reply:
Agree, `snapshot.Usage` is a heavy operation.

We should probably make sure it optional or query based. In the long run, this metrics should be covered by the metrics endpoint proposed in https://github.com/kubernetes/enhancements/issues/2371 based on my understanding.

Reply:
* yes,cadvisor will call `snapshot.Usage` every house keeping period
* the perf impact depends on how many files  there. if there are too many files  ,the cost will be expensive

The metrics that the cadvisor will collect are configurable, but at current , we share a common configuration in kubelet,  how about  using  different configuration  according to the runtime ,because the situation is different between container runtimes .

And for metrics that is already returned by CRI , there is no need to collect it again in cadvisor.

Reply:
Makes sense! To confirm, the idea you're proposing is basically that kubelet will disable calling `snapshot.Usage` (i.e. option #2 in https://github.com/google/cadvisor/pull/2872#discussion_r675855527

Is there a way we can configure cAdvisor in kubelet to disable this collection? If not, do we need add something to cAdvisor to make this configurable?

Reply:
No , the metric list declared in [kubelet's server](https://github.com/kubernetes/kubernetes/blob/fffaadc01331cca57cafa0fc066a2a3eec23acb8/pkg/kubelet/server/server.go#L357-L366) is the metrics that will export to user.  The source of the data is in `pkg/kubelet/stats`,  what we need to change is the data source. 

And cadvisor is configurable :

* [The place where we configure cadvisor in kubelet](https://github.com/kubernetes/kubernetes/blob/fffaadc01331cca57cafa0fc066a2a3eec23acb8/pkg/kubelet/cadvisor/cadvisor_linux.go#L85)
* Where cadvisor acts according to the configuration .
  * [for metrics from cgroup](https://github.com/google/cadvisor/blob/master/container/libcontainer/handler.go#L72)
  *  [for diskUsage metrics](https://github.com/google/cadvisor/blob/master/container/docker/handler.go#L241)

Reply:
Got it. So does it mean that after this change is merged we will want to basically remove `container.DiskUsageMetrics` for containerd [1] in kubelet?

It would be good to understand what specific changes/patch we will need to make in kubelet as part of including this change. Would you be able to maybe draft the patch you're proposing? That way it will be easier to understand what we need to do to integrate these changes into kubelet.  

[1] https://github.com/kubernetes/kubernetes/blob/fffaadc01331cca57cafa0fc066a2a3eec23acb8/pkg/kubelet/cadvisor/cadvisor_linux.go#L85-L93

Reply:
My pleasure ,I will do it later:)

Reply:
Thanks @yyrdl!

Would you be able to draft a PR of the kubelet changes required prior to merging this PR? 

I would just like to ensure we know exactly what changes are required to kubelet first, prior to merging this PR. That way it's clear if we need to make any other cAdvisor/kubelet changes to fully integrate these changes since most important consumer of cAdvisor here will be kubelet.

Reply:
It takes time to redesign the StatsProvider in kubelet , I am a little busy , but I will do it as soon as possible.

Reply:
@yyrdl FYI, I work with @bobbypage and we are solving another cadvisor-containerd issue which needs CRI as well. To work around the circular dependency issue, current plan is integrating cadvisor with CRI in a separate branch. https://github.com/google/cadvisor/pull/2923

Reply:
@qiutongs  good job!

Reply:
@yyrdl I sent you a message on the slack and we can discuss more.